vars/signArtifacts.groovy

```diff
@@ -0,0 +1,25 @@
+void call(Map args = [:]) {
+
+    environment {
+        // These ENV variables are required by https://github.com/opensearch-project/opensearch-signer-client
+        // This client is invoked internally by the sign script.
+        ROLE = "${SIGNER_CLIENT_ROLE}"
+        EXTERNAL_ID = "${SIGNER_CLIENT_EXTERNAL_ID}"
+        UNSIGNED_BUCKET = "${SIGNER_CLIENT_UNSIGNED_BUCKET}"
+        SIGNED_BUCKET = "${SIGNER_CLIENT_SIGNED_BUCKET}"
+    }
+
+    if( !fileExists("$WORKSPACE/sign.sh")) {
```

Comment:
Is the if necessary? Doesn’t it do a `git update` if the code is checked out? Without it you’d get stale files.

Reply:
yes. We would want to ensure to use the same code for the entire signing process.

Reply:
I get what Abhinav is saying as he is trying to prevent updating main branch if any new code is being committed during the run.

Reply:
I am very confused, maybe someone can explain how the `git` call in a Jenkins pipeline work at all? What I really want is that every time we execute the pipeline, it checks out the same opensearch-build code across all pipeline stages. Maybe we need to check it out at a revision that's being built and not just `main` to make it stable?

- If I run the job, then update sign.sh, then run the job again, doesn't that preserve the workspace and isn't the second run using stale code without this update? How do I ever update sign.sh?
- Aren't docker containers always supposed to give me a clean, throw-away environment, in which case this is always `false` when invoked over multiple stages?
- If I have a job that runs in multiple stages in multiple docker containers, am I still getting different code by having `git url: 'https://github.com/opensearch-project/opensearch-build.git', branch: 'main'` with this `if`?

Reply:
- All the jenkins code is run in docker. Anything that is ever created will be deleted after the job completes it's run.
- The condition will not be always in the case of `promoteArtifacts.groovy` as you see, it is repeatedly being called as an action to sign the artifacts before uploading. So, here for example, suppose the job takes an hour to complete, and someone updates the `main`  branch in the meanwhile, it would then start using a different version of code to sign which is not desirable.

Reply:
Isn't it simpler to check out the tooling once in the beginning of the job?

Either way this method won't solve runs across stages.

Reply:
Yes, but checking it out here makes this function self-sufficient to `signArtifacts()` and would not rely on the job calling it to do so. I can look into the commit id var

Reply:
This function is very far from self-sufficient. If you wanted to make it self sufficient, you'd have to apply the same argument further, and also install other tools here, like gpg and check whether they are installed. You're making code harder to maintain by spreading setup all over the place.

Reply:
Well... if we start adding checks for tools installation, no code will ever be self sufficient. 

Trying to understand your point here since setting up the gpg keys is only used by this function here. If we don't do it here, every function would have to call this function anyways before calling `signArtifacts`. I think that make the code and implementation more difficult to maintain. thoughts?

Reply:
First, we can agree to disagree, not a must have for me.

> If we don't do it here, every function would have to call this function anyways before calling signArtifacts.

Every _workflow_ that needs to do signing will need to install the GPG keys, not every function. To me, it's easier to maintain. What you call "self sufficient" sounds like "kitchen sink" to me, the code tries to do too much.

Reply:
Sorry for the terminology there, I meant workflow when I said function there.

Yes, every workflow who would be signing would need to install the gpg keys. Therefore, it would make sense to create the keys inside of signing right? Otherwise, it would be easy to miss creating them for anyone who may use signing in future.

Let's agree to disagree I guess.

Reply:
> Sorry for the terminology there, I meant workflow when I said function there.
> 
> Yes, every workflow who would be signing would need to install the gpg keys. Therefore, it would make sense to create the keys inside of signing right? Otherwise, it would be easy to miss creating them for anyone who may use signing in future.

No, for all the other reasons above.